File: backend/src/test/java/com/codeloom/backend/it/TopicIT.java
package com.codeloom.backend.it;

import static org.hamcrest.Matchers.notNullValue;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

import com.codeloom.backend.dao.topic.TopicRepository;
import com.codeloom.backend.model.Topic;
import java.util.UUID;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.test.context.jdbc.Sql;

@Sql(statements = "TRUNCATE TABLE topics CASCADE", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
class TopicIT extends BackendIntegrationTestSupport {
    @Autowired
    TopicRepository topics;

    private Topic topic(String name) {
        return topics.save(Topic.builder().name(name).build());
    }

    @Nested
    class FindAll {
        @Test
        void empty() throws Exception {
            mockMvc.perform(get("/v1/topics"))
                    .andExpect(status().isOk())
                    .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
                    .andExpect(jsonPath("$.length()").value(0));
        }

        @Test
        void returnsTopics() throws Exception {
            topic("Arrays");
            topic("Dynamic Programming");
            mockMvc.perform(get("/v1/topics"))
                    .andExpect(status().isOk())
                    .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
                    .andExpect(jsonPath("$.length()").value(2))
                    .andExpect(jsonPath("$[0].name").value("Arrays"))
                    .andExpect(jsonPath("$[1].name").value("Dynamic Programming"));
        }
    }

    @Nested
    class FindOne {
        @Test
        void returnsTopic() throws Exception {
            var topic = topic("Graphs");
            mockMvc.perform(get("/v1/topics/{id}", topic.getId()))
                    .andExpect(status().isOk())
                    .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
                    .andExpect(jsonPath("$.id").value(topic.getId().toString()))
                    .andExpect(jsonPath("$.name").value(topic.getName()));
        }

        @Test
        void missingTopicIsNotFound() throws Exception {
            mockMvc.perform(get("/v1/topics/{id}", UUID.randomUUID())).andExpect(status().isNotFound());
        }
    }

    @Nested
    class Create {
        @Test
        void createsTopic() throws Exception {
            mockMvc.perform(post("/v1/topics")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content("{\"name\":\"Greedy\"}"))
                    .andExpect(status().isOk())
                    .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
                    .andExpect(jsonPath("$.id", notNullValue()))
                    .andExpect(jsonPath("$.name").value("Greedy"));
            assertEquals(1, topics.count());
        }

        @Test
        void blankNameIsBadRequest() throws Exception {
            mockMvc.perform(post("/v1/topics")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content("{\"name\":\"\"}"))
                    .andExpect(status().isBadRequest())
                    .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
                    .andExpect(jsonPath("$.status").value(400));
            assertEquals(0, topics.count());
        }
    }

    @Nested
    class Update {
        @Test
        void updatesTopic() throws Exception {
            var topic = topic("Old Name");
            mockMvc.perform(put("/v1/topics/{id}", topic.getId())
                            .contentType(MediaType.APPLICATION_JSON)
                            .content("{\"name\":\"New Name\"}"))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$.id").value(topic.getId().toString()))
                    .andExpect(jsonPath("$.name").value("New Name"));
            assertEquals(
                    "New Name", topics.findById(topic.getId()).orElseThrow().getName());
        }

        @Test
        void rejectsIncompleteUpdate() throws Exception {
            var topic = topic("Old Name");
            mockMvc.perform(put("/v1/topics/{id}", topic.getId())
                            .contentType(MediaType.APPLICATION_JSON)
                            .content("{}"))
                    .andExpect(status().isBadRequest())
                    .andExpect(jsonPath("$.status").value(400));
        }
    }

    @Nested
    class Delete {
        @Test
        void deletesTopic() throws Exception {
            var topic = topic("Bit Manipulation");
            mockMvc.perform(delete("/v1/topics/{id}", topic.getId())).andExpect(status().isOk());
            assertEquals(0, topics.count());
        }
    }
}

File: backend/src/test/kotlin/com/codeloom/backend/it/TopicIT.kt
package com.codeloom.backend.it

import com.codeloom.backend.BaseTest
import com.codeloom.backend.dao.TopicRepository
import com.codeloom.backend.model.Topic
import org.hamcrest.Matchers
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.context.jdbc.Sql
import org.springframework.test.web.servlet.*
import java.util.*
import kotlin.test.assertEquals

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc(addFilters = false)
@Sql(
    statements = [
        "TRUNCATE TABLE topics CASCADE"
    ],
    executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD
)
class TopicIT : BaseTest() {

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Autowired
    private lateinit var topicRepository: TopicRepository

    private fun initTopic(name: String = "Arrays"): Topic =
        topicRepository.save(
            Topic(
                name = name
            )
        )

    /* -------------------------------------------------------------
     * GET /v1/topics
     * ------------------------------------------------------------- */
    @Nested
    inner class FindAll {

        @Test
        fun `test with no topics should return empty array`() {
            mockMvc.get("/v1/topics")
                .andExpect {
                    status { isOk() }
                    content { contentTypeCompatibleWith(MediaType.APPLICATION_JSON) }
                    jsonPath("$.length()", Matchers.equalTo(0))
                }
        }

        @Test
        fun `test with topics should return array`() {
            initTopic("Arrays")
            initTopic("Dynamic Programming")

            mockMvc.get("/v1/topics")
                .andExpect {
                    status { isOk() }
                    content { contentTypeCompatibleWith(MediaType.APPLICATION_JSON) }
                    jsonPath("$.length()", Matchers.equalTo(2))
                    jsonPath("$[0].name", Matchers.equalTo("Arrays"))
                    jsonPath("$[1].name", Matchers.equalTo("Dynamic Programming"))
                }
        }
    }

    /* -------------------------------------------------------------
     * GET /v1/topics/{id}
     * ------------------------------------------------------------- */
    @Nested
    inner class FindOne {

        @Test
        fun `test should return topic`() {
            val topic = initTopic("Graphs")

            mockMvc.get("/v1/topics/${topic.id}")
                .andExpect {
                    status { isOk() }
                    content { contentTypeCompatibleWith(MediaType.APPLICATION_JSON) }
                    jsonPath("$.id", Matchers.equalTo(topic.id.toString()))
                    checkTopic(topic)
                }
        }

        @Test
        fun `test with non-existing id should return 404`() {
            mockMvc.get("/v1/topics/${UUID.randomUUID()}")
                .andExpect {
                    status { isNotFound() }
                }
        }
    }

    /* -------------------------------------------------------------
     * POST /v1/topics
     * ------------------------------------------------------------- */
    @Nested
    inner class Create {

        @Test
        fun `test should create topic`() {
            mockMvc.post("/v1/topics") {
                contentType = MediaType.APPLICATION_JSON
                content = """
                    {
                      "name": "Greedy"
                    }
                """.trimIndent()
            }
                .andExpect {
                    status { isOk() }
                    content { contentTypeCompatibleWith(MediaType.APPLICATION_JSON) }
                    jsonPath("$.id", Matchers.notNullValue())
                    jsonPath("$.name", Matchers.equalTo("Greedy"))
                }

            assertEquals(1, topicRepository.count())
        }

    }

    /* -------------------------------------------------------------
     * PATCH /v1/topics/{id}
     * ------------------------------------------------------------- */
    @Nested
    inner class Patch {

        @Test
        fun `test should patch topic`() {
            val topic = initTopic("Old Name")

            mockMvc.patch("/v1/topics/${topic.id}") {
                contentType = MediaType.APPLICATION_JSON
                content = """
                    {
                      "name": "New Name"
                    }
                """.trimIndent()
            }
                .andExpect {
                    status { isOk() }
                    jsonPath("$.name", Matchers.equalTo("New Name"))
                }

            val updated = topicRepository.findById(topic.id!!).get()
            assertEquals("New Name", updated.name)
        }
    }

    /* -------------------------------------------------------------
     * DELETE /v1/topics/{id}
     * ------------------------------------------------------------- */
    @Nested
    inner class Delete {

        @Test
        fun `test should delete topic`() {
            val topic = initTopic("Bit Manipulation")

            mockMvc.delete("/v1/topics/${topic.id}")
                .andExpect {
                    status { isOk() }
                }

            assertEquals(0, topicRepository.count())
        }
    }

    /* -------------------------------------------------------------
     * JSON assertions
     * ------------------------------------------------------------- */
    private fun MockMvcResultMatchersDsl.checkTopic(topic: Topic) {
        jsonPath("$.id", Matchers.equalTo(topic.id.toString()))
        jsonPath("$.name", Matchers.equalTo(topic.name))
    }
}
